### ISSUE TITLE
hanging requests

### ISSUE THREAD

**Issue by santo74** [arangodb/arangojs#605]
When I start my node.js application in production everything runs fine, but after some time (several hours) I'm starting to get 504 timeouts in the frontend (browser). The 504's are becoming more frequent over time until the whole application is so unstable I have to restart it. (at least once a day).
Obviously a lot of code is involved here (proxies, webservers, etc) but after debugging for more than 2 weeks I can decide that the 504's are always caused by hangs in our DB api.
More specifically, at some point a db request is being fired that never returns. From then on things start to get worse, i.e. more and more db requests will hang and never return resulting in the behaviour explained above.

This is a partial, filtered and redacted, but relevant output of one of our logs:

> server-out.log:2019-05-11T13:43:35.581Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575015580} (api-redacted-1) start query execution
> server-out.log:2019-05-11T13:43:35.583Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575015580} (api-redacted-1) query execution finished
> server-out.log:2019-05-11T13:43:35.584Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575015580} (api-redacted-2) start query execution
> server-out.log:2019-05-11T13:43:35.585Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575015580} (api-redacted-2) query execution finished
> server-out.log:2019-05-11T13:43:35.586Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575015580} (api-redacted-3) start query execution
> server-out.log:2019-05-11T13:43:35.587Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575015580} (api-redacted-3) query execution finished
> server-out.log:2019-05-11T13:43:35.589Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575015580} (api-redacted-4) start query execution
> server-out.log:2019-05-11T13:43:35.592Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575015580} (api-redacted-4) query execution finished
> server-out.log:2019-05-11T13:43:35.593Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575015580} (api-redacted-5) start query execution
> server-out.log:2019-05-11T13:43:35.595Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575015580} (api-redacted-5) query execution finished
> server-out.log:2019-05-11T13:43:35.596Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575015580} (api-redacted-5) start query execution
> server-out.log:2019-05-11T13:43:35.598Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575015580} (api-redacted-5) query execution finished
> server-out.log:2019-05-11T13:43:35.599Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575015580} (api-redacted-6) start query execution
> server-out.log:2019-05-11T13:43:35.601Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575015580} (api-redacted-6) query execution finished
> server-out.log:2019-05-11T13:43:35.602Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575015580} (api-redacted-7) start query execution
> server-out.log:2019-05-11T13:43:35.606Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575015580} (api-redacted-7) query execution finished
> server-out.log:2019-05-11T13:43:35.607Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575015580} (api-redacted-5) start query execution
> server-out.log:2019-05-11T13:43:35.608Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575015580} (api-redacted-5) query execution finished
> server-out.log:2019-05-11T13:43:35.609Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575015580} (api-redacted-1) start query execution
> server-out.log:2019-05-11T13:43:35.609Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575015580} (api-redacted-1) start query execution
> server-out.log:2019-05-11T13:43:35.610Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575015580} (api-redacted-1) start query execution
> server-out.log:2019-05-11T13:43:35.612Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575015580} (api-redacted-1) query execution finished
> server-out.log:2019-05-11T13:43:35.613Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575015580} (api-redacted-1) query execution finished
> server-out.log:2019-05-11T13:43:35.617Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575015580} (api-redacted-8) start query execution
> server-out.log:2019-05-11T13:43:35.618Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575015580} (api-redacted-1) query execution finished
> server-out.log:2019-05-11T13:45:03.302Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-1) start query execution
> server-out.log:2019-05-11T13:45:03.305Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-1) query execution finished
> server-out.log:2019-05-11T13:45:03.305Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-2) start query execution
> server-out.log:2019-05-11T13:45:03.307Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-2) query execution finished
> server-out.log:2019-05-11T13:45:03.307Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-3) start query execution
> server-out.log:2019-05-11T13:45:03.309Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-3) query execution finished
> server-out.log:2019-05-11T13:45:03.311Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-4) start query execution
> server-out.log:2019-05-11T13:45:03.315Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-4) query execution finished
> server-out.log:2019-05-11T13:45:03.317Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-5) start query execution
> server-out.log:2019-05-11T13:45:03.319Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-5) query execution finished
> server-out.log:2019-05-11T13:45:03.319Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-5) start query execution
> server-out.log:2019-05-11T13:45:03.321Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-5) query execution finished
> server-out.log:2019-05-11T13:45:03.322Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-6) start query execution
> server-out.log:2019-05-11T13:45:03.325Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-6) query execution finished
> server-out.log:2019-05-11T13:45:03.326Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-7) start query execution
> server-out.log:2019-05-11T13:45:03.330Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-7) query execution finished
> server-out.log:2019-05-11T13:45:03.331Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-5) start query execution
> server-out.log:2019-05-11T13:45:03.333Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-5) query execution finished
> server-out.log:2019-05-11T13:45:03.334Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-1) start query execution
> server-out.log:2019-05-11T13:45:03.335Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-1) start query execution
> server-out.log:2019-05-11T13:45:03.335Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-1) start query execution
> server-out.log:2019-05-11T13:45:03.339Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-1) query execution finished
> server-out.log:2019-05-11T13:45:03.341Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-1) query execution finished
> server-out.log:2019-05-11T13:45:03.353Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-1) query execution finished
> server-out.log:2019-05-11T13:45:03.363Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-9) start query execution
> server-out.log:2019-05-11T13:45:03.366Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-9) query execution finished
> server-out.log:2019-05-11T13:45:03.367Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-1) start query execution
> server-out.log:2019-05-11T13:45:03.368Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-1) start query execution
> server-out.log:2019-05-11T13:45:03.370Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-1) query execution finished
> server-out.log:2019-05-11T13:45:03.375Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-8) start query execution
> server-out.log:2019-05-11T13:45:03.377Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-1) query execution finished
> server-out.log:2019-05-11T13:45:03.383Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-8) start query execution
> server-out.log:2019-05-11T13:45:03.464Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-8) query execution finished
> server-out.log:2019-05-11T13:45:03.599Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575103301} (api-redacted-8) query execution finished
> server-out.log:2019-05-11T13:49:26.156Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575366155} (api-redacted-1) start query execution
> server-out.log:2019-05-11T13:49:26.159Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575366155} (api-redacted-1) query execution finished
> server-out.log:2019-05-11T13:49:26.160Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575366155} (api-redacted-2) start query execution
> server-out.log:2019-05-11T13:49:26.162Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575366155} (api-redacted-2) query execution finished

For all API calls you can see there is both a "start query execution" and a "query execution finished", except in one case:
server-out.log:2019-05-11T13:43:35.617Z - INFO - [prod-server,2.12.8,arangodb] {request/<hostname-redacted>/1557575015580} (api-redacted-8) start query execution

-> For frontend request {request/<hostname-redacted>/1557575015580} we call "method 8" of our API at some point, but as you can see it starts a db query ("start query execution"), but it never returns (no "query execution finished").

After 30 seconds the frontend receives a 504 timeout for request {request/<hostname-redacted>/1557575015580}.
One minute later (more or less) a new request is received, with internal id {request/<hostname-redacted>/1557575103301}, and this time the same "method 8" of our API is called (even twice) without any issues.

This is what the relevant code looks like:
```
try {
    logger.info(logPrefix + "start query execution");
    a.useDatabase(defaultDb);
    a.query(q, bindVars).then(cursor => {
        logger.info(logPrefix + "query execution finished");
        cursor.all().then(res => {
            return resolve(res);
        })
        .catch((cursorError) => {
            logger.error(logPrefix + "DB cursor Error: %j", cursorError);
            return reject(cursorError);    
        });
    })
    .catch((dbError) => {
        logger.error(logPrefix + "DB Error: %j", dbError);
        return reject(dbError);
    });
}
catch(err) {
    logger.error(logPrefix + "DB Error: %j", err);
    return reject(err);
}
```

As you can see, there's nothing special going on here, we simply call "useDatabase()" followed by "query()" and also catch and log any possible errors.

To be clear, in this case it's "method 8" of our API that hangs, but it can happen with any of our methods, which all use the same code above for executing db queries.

I also don't see any errors on the ArangoDB server and the queries aren't logged as slow queries either.

EDIT:
I'm using arangojs v6.9.0 and ArangoDB v3.3.3 (MMFiles)
I'm planning to switch to a new DB server with ArangoDB v3.4.5 (RocksDB), but I'm having an issue with the file-descriptors limit being too low, cfr [https://github.com/arangodb/arangodb/issues/8965](https://github.com/arangodb/arangodb/issues/8965)

**Comment by pluma** [comment]
This seems to have fallen through the gaps. Does the issue still persist?

Note that invoking `useDatabase` on a database object that's used in other places of your code might lead to unexpected behavior. The upcoming major release will allow creating multiple database objects that share a common connection pool, which may help address any problems this causes.

**Comment by santo74** [comment]
Hi Alan,

The issue doesn't apply anymore, but a lot has happened since I posted this issue so I'm not sure what solved it.
The code I included in the initial post is still the same, but we upgraded ArangoDB to a newer version, switched to the RocksDB engine and even moved to a different hosting provider.
We also rewrote a specific module of our application that was continuously accessing the db (both reads and writes).

Thanks for the info on the upcoming release, I think that might be an interesting improvement.
My guess is that the switch to RocksDB was an important step in resolving the issue but I can't say for sure.
